File: InterviewBit/Strings/palindrome.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <cctype>

using namespace std;

int isPalindrome(string A) {
  vector<char> str;
  for(int itr = 0; itr != A.size(); itr++) {
    if(isalpha(A[itr])) {
      str.push_back(tolower(A[itr]));
    } else if (isdigit(A[itr])) {
      str.push_back(A[itr]);
    }
  }

  if (str.size() == 0)
    return 1;

  for (int begin = 0, end=str.size()-1; begin <= end ; begin++, end--) {
      if (str[begin] != str[end]) {
        return 0;
      }
  }
  return 1;
}

int main() {

  cout<< isPalindrome("1    man, a plan, a canal: Panam1") << endl;

  return 0;
}

File: InterviewBit/Strings/Palindrome.java
package Strings;

/**
 * Given a string, determine if it is a palindrome, considering only alphanumeric characters and ignoring cases.
 * Example:
 * "A man, a plan, a canal: Panama" is a palindrome.
 * "race a car" is not a palindrome.
 * Return 0 / 1 ( 0 for false, 1 for true ) for this problem
 */
public class Palindrome {
    public static void main(String[] args) {
        Solution solution = new Solution();
        int result = solution.isPalindrome("");
        System.out.println(result);
    }

    public static class Solution {
        public int isPalindrome(String A) {
            if (A == null) {
                return 0;
            }
            if (A.length() == 0 || A.length() == 1) {
                return 1;
            }

            A = A.toLowerCase();
            StringBuffer sb = new StringBuffer();

            for (int i = 0; i < A.length(); i++) {
                char ch = A.charAt(i);
                if (Character.isAlphabetic(ch) || Character.isDigit(ch)) {
                    sb.append(ch);
                }
            }

            A = sb.toString();

            int i = 0;
            int j = A.length() - 1;

            while (i < j) {
                if (A.charAt(i) != A.charAt(j)) {
                    return 0;
                }
                i++;
                j--;
            }
            return 1;
        }
    }

}
